File: benchmark/perf_test.dart
// Copyright (c) 2014, Jim Trainor. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

part of benchmark;

Future _runBarePerfTest() {
  const callCount = 1e6;

  dynamic recv(/*ReplyHandler replyHandler,*/ dynamic msg) {
    return msg;
  }

  var jsToDartCompleter = new Completer();

  jsToDartCompletionCallback(elapsedMilliSeconds) {
    var usPerCall = (1.0e6 * elapsedMilliSeconds) ~/ callCount;
    print("bare messaging performance, javascript call to dart = $usPerCall ns/call");
    jsToDartCompleter.complete();
  }
  ;

  bridgeCompletionCallback(elapsedMilliSeconds) {
    var usPerCall = (1.0e6 * elapsedMilliSeconds) ~/ callCount;
    print("js_bridge messaging performance, javascript call to dart = $usPerCall ns/call");
  }
  ;

  js.JsObject perfTest = new js.JsObject(js.context['JsBridgePerfTest'], [recv, callCount, jsToDartCompletionCallback, bridgeCompletionCallback]);

  perfTest.callMethod('runJs2Dart');

  void dartToJsPerfTest(_) {

    var stopWatch = new Stopwatch()..start();
    for (int i = 0; i < callCount; i++) {
      var result = perfTest.callMethod('dart2Js', [i]);
      // don't want overhead of expect call in timed loop, just compare.
      if (result != i) {
        expect("unexpected dart2Js result", isFalse);
      }
    }
    stopWatch.stop();
    var perCall = 1000.0 * stopWatch.elapsedMicroseconds ~/ callCount;
    var units = "ns/call";

    print("bare messaging performance, dart call to javascript = $perCall $units");
  }
  ;

  int dartBenchmarkFunc(int arg) {
    return arg;
  }
  ;

  void bridgePerfTest(_) {
    var bridge = new jsb.JsBridge("benchmark");
    bridge.registerHandler1("dartBenchmarkFunc", dartBenchmarkFunc);
    perfTest.callMethod('runJsBridge2Dart');
  }

  return jsToDartCompleter.future.then(dartToJsPerfTest).then(bridgePerfTest);
}

setupPerfTest() {
  test("js_bridge perf", _runBarePerfTest);
}

File: benchmark/perf_test.js
// Copyright (c) 2014, Jim Trainor. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

/*jslint browser: true, stupid: true */

function JsBridgePerfTest(dartRecvFunc, count, completionCallback, bridgeCompletionCallback) {
    "use strict";
    this.dartRecvFunc = dartRecvFunc;
    this.count = count;
    this.completionCallback = completionCallback;
    this.bridgeCompletionCallback = bridgeCompletionCallback;
}

(function () {
    "use strict";

    // This is the bare js->dart performance benchmark.
    JsBridgePerfTest.prototype.runJs2Dart = function () {
        var i, result, start, stop;

        start = Date.now();
        for (i = 0; i < this.count; i += 1) {
            result = this.dartRecvFunc(i);
            if (result !== i) {
                throw "failed on bad result value " + result + " !== " + this.count;
            }
        }
        stop = Date.now();
        this.completionCallback(stop - start);
    };

    JsBridgePerfTest.prototype.dart2Js = function (msg) {
        return msg;
    };

    // This is called by the dart->js bare call performance test.
    JsBridgePerfTest.prototype.dart2JsBridge = function (msg) {
        return msg;
    };

    // This is the bridged js->dart performance benchmark.
    JsBridgePerfTest.prototype.runJsBridge2Dart = function () {
        var i, result, start, stop,
        dartBenchmarkFunc = JS_BRIDGE.lookupNamespace("benchmark").dartBenchmarkFunc;

        start = Date.now();
        for (i = 0; i < this.count; i += 1) {
            result = dartBenchmarkFunc(i);
            if (result !== i) {
                throw "failed on bad result value " + result + " !== " + this.count;
            }
        }
        stop = Date.now();
        this.bridgeCompletionCallback(stop - start);
    };

}());
